amp.php

```diff
@@ -303,6 +303,24 @@ function amp_deactivate() {
 	flush_rewrite_rules( false );
 }
 
+// @todo Quick fix for now, no centralized flow yet to register services.
+foreach ( [
+	'AmpProject\AmpWP\BackgroundTask\MonitorCssTransientCaching',
+] as $background_task ) {
+	try {
+		( new $background_task() )->register();
+	} catch ( Exception $exception ) {
+		// @todo No logger yet, dump to error log?
+		$exception_class = get_class( $exception );
+		$error_message   = "Exception '{$exception_class}' thrown in background task '{$background_task}': {$exception->getMessage()}";
+		if ( class_exists( 'WP_CLI' ) ) {
+			WP_CLI::warning( $error_message );
+		} else {
+			error_log( $error_message ); // phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_error_log
+		}
+	}
+}
+
```

Comment:
> To follow up with the discussion we just had in the sync call, we should talk about how best to handle this for now: https://github.com/ampproject/amp-wp/pull/4177/files#diff-4805c837d3390517773926fb26726bcdR306-R323

@schlessera I don't really have any better suggestion for the immediate term on where to put this, other than to plop it in `amp.php` like you've done. We'll need to look in the future about how better to organize the `amp.php` bootstrap as part of modularization effort.

The one thing that I think should be done, however, is to avoid calling this right here immediately when the bootstrap is being invoked. This will cause a DB write to occur and potentially race condition on a heavily-trafficked site. This would be adding one more place in addition to the frontend writes being called out in https://github.com/ampproject/amp-wp/issues/3284. 

Since `register_activation_hook()` doesn't work with multisite, what about using `admin_init` instead? This will ensure it only fires once a user has logged in. See:  https://developer.wordpress.org/reference/functions/register_activation_hook/#comment-2100

Reply:
Another option would be to put it inside of an `amp_plugin_update` action, but this still runs during frontend requests. As noted in https://github.com/ampproject/amp-wp/issues/3284#issuecomment-601461083 it could also be moved to `admin_init`. In that way it would be similar to the trigger for the DB upgrade.

Reply:
> This will cause a DB write to occur and potentially race condition on a heavily-trafficked site.

AFAICT, right now, the DB writes are:
- once per day in the `register()` task to schedule a new cron job if none is scheduled;
- once per day in the `process()` that is run via cron.

Am I missing something?

Reply:
Yes, there is the additional write to schedule the interval in the first place. That is, when `register` is called for the very first time upon upgrading, then `wp_next_scheduled()` will return `false` which will then cause `wp_schedule_event()` to be called:

https://github.com/ampproject/amp-wp/blob/d28a5ef1f6e8f930230c357c75c9c30492c60026/src/BackgroundTask/CronBasedBackgroundTask.php#L100-L106

This will then cause `_set_cron_array()` to be called, which will do a DB write. If `wp_schedule_event()` is limited to only happen at `admin_init`, then no cron update would happen on the frontend. When an event transpires, the next event would get scheduled during the cron request, not during a frontend request.

By putting this logic into an `admin_init` hook, this DB write to schedule the event won't happen on the frontend. Granted, this would mean splitting the `register` method into two: one to schedule and one to add the hooks. The schedule part could be done at `admin_init` whereas the hook addition would need to be done with every request (or at least when doing cron).

Maybe I'm being overly cautious about this, and there invariably a lot of other DB writes that happen during frontend requests (especially when transients are used without a persistent object cache). But I'm just trying to think of any ways to make this more robust.

Reply:
I just realised that the schedule call should actually be done on plugin activation instead. That makes it more robust and ensures it is only run once, which is enough, as we are scheduling a recurring event.

Reply:
I've rewritten some parts now to both streamline the way everything is executed and heavily simplify the cron logic.

Reply:
The situation now should be:
- one DB write on plugin activation
- one DB write on plugin deactivation
- one DB write per day during a wp_cron execution

Reply:
Note: The wp_cron one can be filtered out of the frontend processing and only be run as an actual cron background task via filtering and server setup. This should solve the problem completely for high-traffic sites.

Reply:
The plugin activation hook does work on multisite. However, it will only schedule the cron event for the primary site. I can add code to cycle through the sites and schedule an event for each of them.

Reply:
The above is the case if you network-activate the plugin. We could also disallow network-activation, so you'll have to activate on a subsite by subsite basis instead...

Reply:
Here's some code that apparently handles the network activation case and the multisite blog creation case: https://wordpress.stackexchange.com/a/181150

I am not an authority on this!

Reply:
cc @felixarntz!

Reply:
> We could also disallow network-activation, so you'll have to activate on a subsite by subsite basis instead...

No, we should still allow it to be network-activated.

Reply:
@westonruter @schlessera In order to make a site-wide action like this "reliable" for multisite it needs to happen on a hook like `admin_init` (or if earlier is preferable something like `init` with `is_admin()` check).

It's a pain, but the activation hooks don't cascade to the per-site level in multisite during network activation (scale is an issue either way), and they also don't fire when a new site is created if the plugin is already network-active. So using an as-inexpensive-as-possible DB call to check whether the action still needs to be done (only in admin!) would be my recommendation for now.

Reply:
@felixarntz But that will inevitably run into race conditions then, and we'll end up running the cron job multiple times per day, potentially disabling the transient caching much too soon.

Reply:
Note: In this case, not running the cron job at all is highly preferable to running it multiple times because of a race condition.

Reply:
Why would it not be reliable to cycle over all blogs (once!) on plugin activation (something that is under the control of the administrator, and listening to site creation?

Reply:
@schlessera Potentially yes, but we can reduce the chance for this to a minimum, e.g. we could only schedule the cron job if an administrator is logged in. I would say it's highly unlikely that a site has so many administrators that more than one loads an admin page in the same given second right in the moment where this action is executed.

Reply:
@schlessera 
> Why would it not be reliable to cycle over all blogs (once!) on plugin activation (something that is under the control of the administrator, and listening to site creation?

Yes we could do that but then you'll run into scaling issues with larger multisites.